### PR TITLE
feat!: Refactor module to use maps instead of lists

### LINKED ISSUE #289: Putting listener rules in the middle of the list causes erratic terraform behavior
## Description

When adding a new rule to the middle of https_listeners array after creating the stack, Terraform fails to update the listener rule with a new ID, causing duplicate rules with the same ID. 

Steps to Reproduce:

* Apply the  attached Terraform plan containing the initial listener rules. (You need to provide your own certificate ARN)
* Uncomment the existing rule in the middle of the http_tcp_listeners or https_listeners array, marked as such
* Apply the updated plan.
* Notice the duplicate rule with the same ID is created.
* Attempt to apply the plan again without any changes to the Terraform code and notice Terraform overwriting existing rule and never converging 

Expected Results:

When adding a new rule to the middle of the http_tcp_listeners or https_listeners array, Terraform should update the listener rule with a new ID, rather than creating a duplicate rule with the same ID.

- [X] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: 8.6.0

- Terraform version:
1.3.8
1.4.4

- Provider version(s):

registry.terraform.io/hashicorp/aws v4.62.0

## Reproduction Code [Required]

```
provider "aws" {
  region = local.region
}

data "aws_availability_zones" "available" {}

locals {
  name   = "ex-${basename(path.cwd)}"
  region = "us-east-1"

  vpc_cidr = "10.0.0.0/20"
  azs      = slice(data.aws_availability_zones.available.names, 0, 2)

}

##################################################################
# Application Load Balancer
##################################################################

module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 8.6"

  load_balancer_type = "application"
  internal           = true

  name    = "my-alb-int"
  vpc_id  = module.vpc.vpc_id
  subnets = module.vpc.private_subnets

  security_groups = [module.alb_sg.this_security_group_id]

  https_listeners = [
    {
      port            = 443
      protocol        = "HTTPS"
      certificate_arn = var.certificate
      action_type     = "fixed-response"
      fixed_response = {
        content_type = "text/plain"
        status_code  = "200"
        message_body = "Nothing to see here move along"
      }
    },
  ]

  https_listener_rules = [
    {
      https_listener_index = 0
      priority             = 100

      actions = [{
        type               = "forward"
        target_group_index = 0
      }]

      conditions = [{ host_headers = ["my-host-header.example.com"] }]
    },
    # UNCOMMENT THIS RULE AFTER APPLYING THE PLAN FOR THE FIRST TIME
    # {
    #   https_listener_index = 0
    #   priority             = 102

    #   actions = [{
    #     type               = "forward"
    #     target_group_index = 0
    #   }]

    #   conditions = [{ host_headers = ["my-host-header2.example.com"] }]
    # },
    {
      https_listener_index = 0
      priority             = 114

      actions = [{
        type               = "forward"
        target_group_index = 0
      }]

      conditions = [{ host_headers = ["my-host-header3.example.com"] }]
    },
  ]

  target_groups = [
    {
      name_prefix          = "my-tg"
      backend_protocol     = "HTTPS"
      backend_port         = 443
      target_type          = "ip"
      deregistration_delay = 60
      health_check = {
        path                = "/"
        port                = "traffic-port"
        protocol            = "HTTPS"
        healthy_threshold   = 2
        unhealthy_threshold = 2
        timeout             = 5
        interval            = 30
      }
    }
  ]

}




################################################################################
# Supporting resources
################################################################################

module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "~> 3.0"

  name = local.name
  cidr = local.vpc_cidr

  azs             = local.azs
  private_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 4, k)]
  public_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 48)]

  enable_nat_gateway   = true
  single_nat_gateway   = true
  enable_dns_hostnames = true


}

module "alb_sg" {

  source  = "terraform-aws-modules/security-group/aws"
  version = "3.0.0"

  name        = "alb-sg"
  description = "Security group for the ALB"
  vpc_id      = module.vpc.vpc_id

}


variable "certificate" {
  default = "INSERT YOUR CERTIFICATE ARN HERE"
  description = "Provide the certificate ARN"
}

```

No

Yes

1.  Set the value of the 'certificate' variable to your own certificate
2.  Apply the terraform plan to create initial stack,  it will succeed
3.  Uncomment the lines containing the rule in the middle of the array

## Expected behavior

A rule added to the listener when the stack is applied with it's own unique ID

## Actual behavior

The new rule has the same ID as the rule that was in that position,  this causes each run of terraform to overwrite the rule definition either with the previous rule or the one you just added,  the stack never converges

### LINKED ISSUE #298: Associate WAF variable
## Is your request related to a new offering from AWS?
Is this functionality available in the [AWS provider for Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs)? See [CHANGELOG.md](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md), too.
  - No 🛑: please wait to file a request until the functionality is avaialble in the AWS provider
  - Yes ✅: please list the AWS provider version which introduced this functionality

## Is your request related to a problem? Please describe.
I am working on updating to use a WAF on an existing LB. However, if the WAF is not already created I get the invalid count error.

```
846:   count        = var.web_acl_arn != null ? 1 : 0

The "count" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many instances will be created. To work
around this, use the -target argument to first apply only the resources that the count depends on.
```

## Describe the solution you'd like.
I would like to see if it would be within standards for this project to switch to an "associate_waf" variable similar to the "create_*" variables in this and other aws modules. The reason for this is so that count is no longer dependent on a value and therefore not cause multiple plan and applies.

## Describe alternatives you've considered.
I have not been able to think of anything beyond not using the module which I would like to avoid.

### LINKED ISSUE #299: Weighted target groups don't work
## Description

The current way of defining `target_group_arn` doesn't allow the use of multiple weighted target groups. 

As per the [`lb_listener#target_group_arn`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener#target_group_arn) documentation:

> `target_group_arn` - (**Optional**) ARN of the Target Group to which to route traffic. Specify only if `type` is `forward` and you want to route to a **single target group**. To route to one or more target groups, use a `forward` block instead.

> `forward` - (...) If you specify both `forward` block and `target_group_arn` attribute, you can specify only one target group using `forward` and it must be the same target group specified in `target_group_arn`.

The way the `target_group_arn` is currently defined means that this value will always have a value if the `action_type` is `forward` (default). This means that when using multiple target groups in the forward block, the single `target_group_arn` will take precedence. 

https://github.com/terraform-aws-modules/terraform-aws-alb/blob/cb8e43d456a863e954f6b97a4a821f41d4280ab8/main.tf#L638

This was previously [mentioned](https://github.com/terraform-aws-modules/terraform-aws-alb/issues/280#issuecomment-1556389320) in this comment in a related issue

## Possible fix

Make the `target_group_arn` condition also check if there's more than one target group. I tested this in the terraform console with this expression

```hcl
length(lookup(lookup(default_action.value, "forward", {}), "target_groups", []))
```
With the full expression being something like this:

```hcl
target_group_arn = contains([null, "", "forward"], lookup(default_action.value, "action_type", "")) && length(lookup(lookup(default_action.value, "forward", {}), "target_groups", [])) < 2 ? aws_lb_target_group.main[lookup(default_action.value, "target_group_index", count.index)].id : null
```

## Versions

- Module version [Required]: `v8.7.0`
- Terraform version:
```
terraform -version
Terraform v1.3.4
on darwin_arm64
```
- Provider version(s):
```
Terraform v1.3.4
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v5.10.0
```

## Reproduction Code [Required]

I believe the example code should be enough to reproduce this
https://github.com/terraform-aws-modules/terraform-aws-alb/blob/cb8e43d456a863e954f6b97a4a821f41d4280ab8/examples/complete-alb/main.tf#L82-L98

1. Define more than one target groups
2. Define an `http_tcp_listeners` block with multiple weighted `target_groups`
3. See terminal output and screenshots below

## Expected behavior

When more than one `target_groups` is defined, `target_group_arn` should be `null`

## Actual behavior

When more than one `target_groups` is defined, `target_group_arn` is still defined, essentially making the listener rule forward traffic to a single target group. 

It's worth mentioning that Terraform doesn't error

### Terminal Output Screenshot(s)

This is how my terraform plan looks like. Notice how both the `target_group_arn` and the multiple `target_group`s are defined
```
      + default_action {
          + target_group_arn = "arn:aws:elasticloadbalancing:us-east-1:012345678912:targetgroup/MY-TARGET-GROUP-1/x0x0x0x0x0x0x"

          + forward {
              + stickiness {
                  + duration = 3600
                  + enabled  = true
                }

              + target_group {
                  + arn    = "arn:aws:elasticloadbalancing:us-east-1:012345678912:targetgroup/MY-TARGET-GROUP-1/x0x0x0x0x0x0x"
                  + weight = 100
                }
              + target_group {
                  + arn    = "arn:aws:elasticloadbalancing:us-east-1:012345678912:targetgroup/MY-TARGET-GROUP-2/x0x0x0x0x0x0x"
                  + weight = 0
                }
            }
        }
```

This results in the target group looking like this

image

When the expected behavior is for it to look like this

image

### LINKED ISSUE #301: getting Invalid index while crearing alb with huge #  of listener rules
## Description
When the alb is run with  a huge # of listener rules specially, it fails with these errors. 

Error: Invalid index
│
│   on ../terraform-aws-alb/main.tf line 285, in resource "aws_lb_listener_rule" "https_listener_rule":
│  285:       target_group_arn = aws_lb_target_group.main[lookup(action.value, "target_group_index", count.index)].id
│     ├────────────────
│     │ action.value is object with 3 attributes
│     │ aws_lb_target_group.main is tuple with 45 elements
│     │ count.index is 52
│
│ The given key does not identify an element in this collection value.
╵
I am using the input file as in example, defined https_listeners[]. http_tcp_listeners[], and their respective listener rules, and target as a list of map target_group[ [{},{}]

### LINKED ISSUE #304: Support for network load balancer (NLB) security groups
Official announcement from AWS can be fond [here](https://aws.amazon.com/about-aws/whats-new/2023/08/network-load-balancer-supports-security-groups/).

The [functionality](https://github.com/hashicorp/terraform-provider-aws/issues/32982) has been added to the AWS provider in [v5.13](https://github.com/hashicorp/terraform-provider-aws/releases/tag/v5.13.0) if I understood that right. At least something about this new feature can be found in the release notes.

While https://github.com/hashicorp/terraform-provider-aws/pull/33008 states that no implementation updates were need. So there is a good possibility it already should work with pre-v5.13 releases of the provider.

### PR DESCRIPTION
**Pull request by bryantbiggs**
### Backwards incompatible changes

- `target_groups` previously were defined by an array of target group definitions that were created using the `count` meta-argument. This has been replaced with a map of target group definitions that are created using the `for_each` meta-argument in order to provide better stability when adding/removing target group definitions.
- `target_groups` no longer support multiple targets per target group. There are alternate methods to achieve similar functionality such as weighted target groups or using an autoscaling group as a target when targetting EC2 instances.
- The previous methods for creating listeners have been removed in favor of one argument, `listeners`, which take a map of listener definitions that are created using the `for_each` meta-argument in order to provide better stability when adding/removing listener definitions. Previously the `target_group_index` was used to associate/reference a target group; that is now replaced with `target_group_key` which is the key of the target group definition in the `target_groups` map.
- `security_group_rules` has been replaced by `security_group_ingress_rules` and `security_group_egress_rules` to align with the new underlying resources.
-
- Minimum supported version of Terraform AWS provider updated to `v5.13` to support the latest features provided via the resources utilized.
- Minimum supported version of Terraform updated to `v1.0`
- The `Name` tag has been removed from resources

### Added

- Security group attachment restrictions have been removed now that both ALB and NLB support security groups
- Support for creating Route53 records for ALB/NLB DNS names via the `route53_records` variable

### Modified

- `enable_cross_zone_load_balancing` now defaults to `true`
- `drop_invalid_header_fields` now defaults to `true`
- `enable_deletion_protection` now defaults to `true`
- `associate_web_acl` has been added to identify when a WAFv2 Web ACL should be associated with the ALB; previously this was accomplished by checking for the presence of a value passed to `web_acl_arn` which is known to cause issues when the value does not exist and is computed.

See [UPGRADE-9.0.md](https://github.com/bryantbiggs/terraform-aws-alb/blob/refactor/for-each/UPGRADE-9.0.md) for further details

## Motivation and Context

- Resolves #289
- Resolves #297
- Resolves #298
- Resolves #299
- Resolves #301
- Resolves #304
- Resolves #305
- Resolves #310

## Breaking Changes
- Yes

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by antonbabenko** [comment]
Nice! Let me know when this is ready for my review :)

**Comment by bryantbiggs** [comment]
yes, sorry forgot to hit the button there 😅 - but its ready for review

File: main.tf [hunk 1]
```diff
@@ -1,806 +1,548 @@
...
-          scope                               = lookup(authenticate_cognito.value, "scope", null)
-          session_cookie_name                 = lookup(authenticate_cognito.value, "session_cookie_name", null)
-          session_timeout                     = lookup(authenticate_cognito.value, "session_timeout", null)
-          user_pool_arn                       = authenticate_cognito.value["user_pool_arn"]
-          user_pool_client_id                 = authenticate_cognito.value["user_pool_client_id"]
-          user_pool_domain                    = authenticate_cognito.value["user_pool_domain"]
-        }
-      }
+  target_group_arn  = aws_lb_target_group.this[each.key].arn
+  target_id         = each.value.target_id
```
Comment (ChristopheBougere) [inline main.tf:502]: When target group is used by ECS, target_type is ip but there is no target_id, this line fails. I think you should only create a `aws_lb_target_group_attachment` when there is a `each.value.target_id`, not sure about the right way to do that. I was able to deploy by changing the condition in the `for_each`:

```tf
  for_each = { for k, v in var.target_groups : k => v if false }
```

**Review by ChristopheBougere — COMMENTED** [review]
I'm facing https://github.com/terraform-aws-modules/terraform-aws-alb/issues/289 so I gave a try to your MR.

File: main.tf (hunk 1 above)
Comment (bryantbiggs) [reply main.tf:502]: If I understand correctly, you want to create a target group but not specify what is attached to that target group?

Comment (ChristopheBougere) [reply main.tf:502]: Correct
It's very similar to this example with ALB and ECS on `terraform-aws-ecs`
https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/master/examples/fargate/main.tf
image

The attachment is managed by the autoscaling, so I guess we dont want to specify any ID in terraform

Comment (bryantbiggs) [reply main.tf:502]: this should be resolved in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/d112a97df83035ad9d5cbeda9139b554dba8b740

Comment (ChristopheBougere) [reply main.tf:502]: It works, thanks a lot!

**Comment by spr-mweber3** [comment]
@antonbabenko What do you think? Is this ready to be merged?

**Comment by lblazewski** [comment]
Highly interested in this refactor too!

File: examples/complete-alb/main.tf [hunk 2]
```diff
@@ -506,22 +433,21 @@ resource "null_resource" "download_package" {
 
 module "lambda_with_allowed_triggers" {
   source  = "terraform-aws-modules/lambda/aws"
-  version = "~> 5.0"
+  version = "~> 3.0"
```
Comment (antonbabenko) [inline examples/complete-alb/main.tf:436]:
```suggestion
  version = "~> 6.0"
```

File: examples/complete-alb/main.tf [hunk 3]
```diff
@@ -530,15 +456,14 @@ module "lambda_with_allowed_triggers" {
 
 module "lambda_without_allowed_triggers" {
   source  = "terraform-aws-modules/lambda/aws"
-  version = "~> 5.0"
+  version = "~> 3.0"
```
Comment (antonbabenko) [inline examples/complete-alb/main.tf:459]:
```suggestion
  version = "~> 6.0"
```

File: examples/complete-alb/main.tf [hunk 4]
```diff
@@ -563,9 +488,8 @@ module "vpc" {
   private_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 4, k)]
   public_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 48)]
 
-  # Disabled NAT gateway to save a few seconds running this example
-  enable_nat_gateway   = false
-  enable_dns_hostnames = true
+  enable_nat_gateway = true
```
Comment (antonbabenko) [inline examples/complete-alb/main.tf:491]:
```suggestion
  enable_nat_gateway = false
```

I believe that we don't need NAT in examples.

File: examples/complete-alb/main.tf [hunk 5]
```diff
@@ -576,20 +500,30 @@ data "aws_route53_zone" "this" {
 
 module "acm" {
   source  = "terraform-aws-modules/acm/aws"
-  version = "~> 4.0"
+  version = "~> 3.0"
```
Comment (antonbabenko) [inline examples/complete-alb/main.tf:503]:
```suggestion
  version = "~> 4.0"
```

File: examples/complete-alb/main.tf [hunk 6]
```diff
@@ -576,20 +500,30 @@ data "aws_route53_zone" "this" {
...
 
   domain_name = var.domain_name
   zone_id     = data.aws_route53_zone.this.id
 }
 
 module "wildcard_cert" {
   source  = "terraform-aws-modules/acm/aws"
-  version = "~> 4.0"
+  version = "~> 3.0"
```
Comment (antonbabenko) [inline examples/complete-alb/main.tf:511]:
```suggestion
  version = "~> 4.0"
```

**Review by antonbabenko — CHANGES_REQUESTED** [review]
It looks great, mostly small questions, but also some thoughts about simplifying the map value (WDYT?).

File: README.md [hunk 7]
```diff
@@ -217,69 +250,87 @@ When you're using ALB Listener rules, make sure that every rule's `actions` bloc
...
 }
 ```
 
-## Assumptions
-
-It's recommended you use this module with [terraform-aws-vpc](https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws), [terraform-aws-security-group](https://registry.terraform.io/modules/terraform-aws-modules/security-group/aws), and [terraform-aws-autoscaling](https://registry.terraform.io/modules/terraform-aws-modules/autoscaling/aws/).
-
-## Notes
-
-1. Terraform AWS provider version v2.39.0 and newer has [issue #16674](https://github.com/hashicorp/terraform-provider-aws/issues/16674) related to "Provider produced inconsistent final plan". It means that S3 bucket has to be created before referencing it as an argument inside `access_logs = { bucket = "my-already-created-bucket-for-logs" }`, so this won't work: `access_logs = { bucket = module.log_bucket.s3_bucket_id }`.
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply README.md:271]: Re-enabling this in the example worked after the changes so I assumed it was now resolved - not 100% certain but now "it works" 🤷🏽 😅

File: main.tf [hunk 8]
```diff
@@ -809,41 +551,79 @@ resource "aws_security_group" "this" {
...
+  tags = merge(var.tags, var.security_group_tags)
 
   lifecycle {
     create_before_destroy = true
   }
 }
 
-resource "aws_security_group_rule" "this" {
-  for_each = { for k, v in var.security_group_rules : k => v if local.create_security_group }
+resource "aws_vpc_security_group_egress_rule" "this" {
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply main.tf:564]: good point, I'll add. For the difference, see the note on the docs here https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule.html

image

File: examples/complete-alb/main.tf (hunk 6 above)
Comment (bryantbiggs) [reply examples/complete-alb/main.tf]: fixed in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: examples/complete-alb/main.tf (hunk 5 above)
Comment (bryantbiggs) [reply examples/complete-alb/main.tf]: fixed in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: examples/complete-alb/main.tf (hunk 4 above)
Comment (bryantbiggs) [reply examples/complete-alb/main.tf]: fixed in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: examples/complete-alb/main.tf (hunk 3 above)
Comment (bryantbiggs) [reply examples/complete-alb/main.tf]: fixed in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: examples/complete-alb/main.tf (hunk 2 above)
Comment (bryantbiggs) [reply examples/complete-alb/main.tf]: fixed in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: main.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply main.tf:564]: added in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: UPGRADE-9.0.md [hunk 9]
```diff
@@ -0,0 +1,892 @@
...
+```
+
+## Terraform State Moves
+
+### Listener(s)
+
+Each listener will need to be migrated using the index position in v8.x to the key of the listener in v9.x - see example state move commands below:
+
+```sh
+tf state mv 'module.alb.aws_lb_listener.frontend_http_tcp[0]' 'module.alb.aws_lb_listener.this["default"]'
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply UPGRADE-9.0.md]: whoops 😅  - fixed in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/522c07009fa3cc79200886990f3ab8ba1ac76231

File: UPGRADE-9.0.md [hunk 10]
```diff
@@ -0,0 +1,892 @@
...
+      certificate_arn             = module.acm.acm_certificate_arn
+      additional_certificate_arns = [module.wildcard_cert.acm_certificate_arn]
+
+      forward = {
+        target_group_key = "lambda-with-trigger"
+      }
+
+      rules = {
+        cognito = {
+          actions = [
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply UPGRADE-9.0.md:622]: yes - any of the networking resources tend to get very gnarly with just miles of nested maps. I actually created an AppMesh module awhile ago but after seeing such deeply nested maps everywhere, I felt like that wasn't going to do any good so it never saw the light of day 😅 (see snippet below)

I think this starts getting into the specific examples/blueprints territory - a few questions for you:
1. Do we want to show actual examples or just snippets in docs
2. Should that be added in this PR or follow up

Ugly, deeply nested maps 🙈 
image

**Comment by bryantbiggs** [comment]
also, do you know whats up with the wrappers? I can't figure out why the CI keeps showing a diff, even if I manually fix it

**Comment by spr-mweber3** [comment]
@antonbabenko Looks good to me.

**Comment by LeszekBlazewski** [comment]
Waiting for this feature too!

**Comment by LeszekBlazewski** [comment]
Any update on this PR? @antonbabenko 

I think the majority of the work has been done and only few tweaks are needed to push this through. I think many consumers of the module can benefit from it!

File: UPGRADE-9.0.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (antonbabenko) [reply UPGRADE-9.0.md:622]: Nested maps are, unfortunately, the best solution. We can help users by showing snippets in the docs.

Let's update README in the root module and show how to use this module "in general for ALB" and include multiple small chunks of configurations to illustrate different "actions" (forward, redirect, fixed-response, etc).

File: UPGRADE-9.0.md [hunk 11]
```diff
@@ -0,0 +1,931 @@
...
+Each lambda permission will need to be migrated using the <index.key> position in v8.x to the key of the lambda permission in v9.x - see example state move commands below:
+
+```sh
+terraform state mv 'module.alb.aws_lambda_permission.lb["2.lambda_without_allowed_triggers"]' 'module.alb.aws_lambda_permission.this["lambda-without-trigger"]'
+```
+
+### Security Group Rule(s)
+
+The security group rules have been changed from the `aws_security_group_rule` resource to the new `aws_vpc_security_group_ingress_rule`/`aws_vpc_security_group_egress_rule` resources.
+If you do not wish for the rules to be recreated during the upgrade, you will need to remove the existing rules from the Terraform state, and re-import (you cannot `terrraform mv` across
```
Comment (antonbabenko) [inline UPGRADE-9.0.md:896]:
```suggestion
If you do not wish for the rules to be recreated during the upgrade, you will need to remove the existing rules from the Terraform state, and re-import (you cannot `terrraform state mv` across
```

File: UPGRADE-9.0.md [hunk 12]
```diff
@@ -0,0 +1,931 @@
...
+different resource types). For example, for one rule you would perform the following given the following code snippets:
```
Comment (antonbabenko) [inline UPGRADE-9.0.md:897]:
```suggestion
different resource types). For example, for one rule, you would perform the following snippet:
```

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/antonbabenko/pre-commit-terraform
-    rev: v1.81.0
+    rev: v1.83.2
```
Comment (antonbabenko) [inline .pre-commit-config.yaml:3]: Please upgrade this version to make sure that ordering in `wrappers` is ok - ref. https://github.com/antonbabenko/pre-commit-terraform/pull/565

**Review by antonbabenko — APPROVED** [review]
Looks very good (and sorry for such a long delay with the review)!

File: .pre-commit-config.yaml (hunk 13 above)
Comment (bryantbiggs) [reply .pre-commit-config.yaml]: updated in https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305/commits/c016541dfe0e4249edcf5d358528853f78a3106e

**Comment by spr-mweber3** [comment]
Almost at the finish line. Love you guys!

File: UPGRADE-9.0.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply UPGRADE-9.0.md:622]: I've updated the main README to show some examples and then I added this https://github.com/bryantbiggs/terraform-aws-alb/blob/refactor/for-each/docs/patterns.md to start collecting contributions on usage patterns from the community. I'll open an issue to signal this and see if we get an Hacktoberfest contributors to these usage patterns! (ALBs and NLBs aren't my wheelhouse so, struggling a bit in the usage pattern area here 😅 )

**Comment by bryantbiggs** [comment]
Not sure if CI will correct itself once this is merged, but I had to manually adjust the wrappers to get it to pass CI. I'll follow up with an issue to ask for contributions to the patterns doc to collect common configuration/usage patterns to make groking the nested maps easier for folks. For now, I think we're close enough so here we go 🚀

**Comment by antonbabenko** [comment]
@bryantbiggs I have just run the hook on this PR locally, and it behaves as on CI. Maybe you have the wrong versions of `hcledit` on your side? (blind guess)

**Review by antonbabenko — APPROVED** [review]
LGTM 🚀

**Comment by bryantbiggs** [comment]
hmm, this is what I have which is the latest 
image

**Comment by antonbabenko** [comment]
This PR is included in version 9.0.0 :tada:

**Comment by bryantbiggs** [comment]
For all following this PR, or those who may find there way here - please come help us contribute the configuration snippets that demonstrate the common usage patterns under the new v9.0 module definition. Details can be found here, thanks all! https://github.com/terraform-aws-modules/terraform-aws-alb/issues/311

File: main.tf (hunk 1 above)
Comment (ChristopheBougere) [reply main.tf:503]: Hi @bryantbiggs , sorry to reopen but it seems https://github.com/terraform-aws-modules/terraform-aws-alb/commit/d112a97df83035ad9d5cbeda9139b554dba8b740 has disappeared in the final MR
Should I open a new issue ?

Comment (bryantbiggs) [reply main.tf:503]: the functionality is still there, just different due to stability reasons - see https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/736a7da9d4a5a514b1196c9b068050b11277b9a7/examples/fargate/main.tf#L237

Comment (ChristopheBougere) [reply main.tf:503]: Oh my mistake thanks!

**Pull request merged**